<a href="https://colab.research.google.com/github/madhuselvaraj/tensorflow_projects/blob/master/mnist_using_conv2D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Here we are going to use convolution of single layer to improve our training accuracy. we can consider images from fashion_mnist to proceed.

At first,import tensorflow

In [0]:
import tensorflow as tf

define mycallback() incase of it reaches an accuracy 99.8% inorder to stop the training

In [0]:
class myCallback(tf.keras.callbacks.Callback):
  def on_epoch_end(self, epoch, logs={}):
    if(logs.get('acc')>0.998):
      print("\nReached 0.998 acc so cancelling training!")
      self.model.stop_training = True
      

In [0]:
callbacks = myCallback()  

downloading images as API from fashion_mnist and loading it as test and training images

In [0]:
mnist = tf.keras.datasets.mnist
(training_images, training_labels), (test_images, test_labels) = mnist.load_data()

reshaping our model for implementing convolution layer.(i.e) we cannot give the exact training set to conv2d and so reshaping it as we give for flatten()

In [0]:
training_images=training_images.reshape(60000, 28, 28, 1)
training_images = training_images/255
test_images=test_images.reshape(10000, 28, 28, 1)
test_images = test_images/255

defining our model with conv2D 64 filters 

In [6]:

model = tf.keras.models.Sequential([
                                  tf.keras.layers.Conv2D(64,(3,3),activation = 'relu',input_shape = (28,28,1)),
                                  tf.keras.layers.MaxPooling2D(2, 2),
                                  tf.keras.layers.Flatten(),
                                   tf.keras.layers.Dense(128,activation = tf.nn.relu),                                   
                                    tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

Instructions for updating:
Colocations handled automatically by placer.


fit our training data and evaluating out test set

In [7]:
model.fit(training_images, training_labels, epochs=20,callbacks=[callbacks])

test_loss = model.evaluate(test_images, test_labels)

Epoch 1/20
60000/60000 [==============================] - 81s 1ms/sample - loss: 0.1426 - acc: 0.9566
Epoch 2/20
60000/60000 [==============================] - 82s 1ms/sample - loss: 0.0485 - acc: 0.9851
Epoch 3/20
60000/60000 [==============================] - 83s 1ms/sample - loss: 0.0305 - acc: 0.9908
Epoch 4/20
60000/60000 [==============================] - 83s 1ms/sample - loss: 0.0196 - acc: 0.9934
Epoch 5/20
60000/60000 [==============================] - 84s 1ms/sample - loss: 0.0125 - acc: 0.9961
Epoch 6/20
60000/60000 [==============================] - 83s 1ms/sample - loss: 0.0098 - acc: 0.9968
Epoch 7/20
60000/60000 [==============================] - 85s 1ms/sample - loss: 0.0083 - acc: 0.9973
Epoch 8/20
59968/60000 [============================>.] - ETA: 0s - loss: 0.0062 - acc: 0.9980
Reached 0.998 acc so cancelling training!
10000/10000 [==============================] - 3s 312us/sample - loss: 0.0493 - acc: 0.9862


Here is our results. Training accuracy is about 99.8% and training loss is 0.0062.
Test accuracy is about 98.62% and loss about 0.493 which is not bad i think!